graph_algorithms/algorithms/examples/louvain.gsql

```diff
@@ -1,80 +1,295 @@
-                                      @@edgeCntIncidentCluster += (s.@cluster->1)
-                              ELSE
-                                      @@edgeCntWithinCluster += (s.@cluster->1)
-                              END
-                        POST-ACCUM
-                                INT resultCluster = s.@cluster,
-                                INT m = @@totalNumEdges,
-                                FLOAT maxDeltaQ = -999,
-                                FLOAT deltaQ_new = 0,
-                                FOREACH (cluster, edgeCnt) IN s.@clusterEdgeCnt DO
-                                        INT within = @@edgeCntWithinCluster.get(cluster),
-                                        INT incident = @@edgeCntIncidentCluster.get(cluster),
-                                        deltaQ_new = ((within + s.@clusterEdgeCnt.get(cluster)) / (2.0*m))
-                                                     - pow((incident + s.@edgeCnt)/(2.0*m),2)
-                                                     - (within / (2.0*m)
-                                                     - pow((incident/(2.0*m)),2)
-                                                     - pow((s.@edgeCnt/(2.0*m)),2)),
-                                        IF deltaQ_new > maxDeltaQ AND deltaQ_new > s.@deltaQ AND deltaQ_new > 0 THEN
-                                                maxDeltaQ = deltaQ_new,
-                                                resultCluster = cluster
-                                        END
-                                END,
-                                IF s.@cluster != resultCluster THEN
-                                        s.@cluster = resultCluster,
-                                        s.@deltaQ = maxDeltaQ,
-                                        @@changed += true
-                                END,
-                                s.@clusterEdgeCnt.clear()
-                        ;
-                @@edgeCntWithinCluster.clear();
-                @@edgeCntIncidentCluster.clear();
+        S = SELECT s 
+            FROM Start:s
+            ACCUM @@totIncidentCluster += (s.@cid -> s.@weight); 
+              
+        iteration = 1;
+        Iter1 = iter1 - 1;
+              
+        WHILE (iteration < 2 OR @@modularity - last_modularity > epsilon) LIMIT Iter1 DO
+                iteration += 1;
+                loop = 0;
+                WHILE (loop < partitions) DO 
+                        S = SELECT s 
+                            FROM Start:s -(Friend:e)-> :t
+                            WHERE s.@uid % partitions == loop AND abs(s.@weight - @@totIncidentCluster.get(s.@cid)) > epsilon OR abs(t.@weight - @@totIncidentCluster.get(t.@cid)) > epsilon OR (abs(s.@weight - @@totIncidentCluster.get(s.@cid)) < epsilon AND abs(t.@weight - @@totIncidentCluster.get(t.@cid)) < epsilon AND s.@cid > t.@cid)   # at least one cluster not only itself, or use smaller label
```

Comment:
And what if this is a directed graph A->B->C?

Reply:
louvain supports only undirected graph